README.md
[![codecov](https://codecov.io/gh/DeepSpace2/StyleFrame/branch/master/graph/badge.svg?token=f4Q048ZOwC)](https://codecov.io/gh/DeepSpace2/StyleFrame)
[![GitHub Actions](https://img.shields.io/github/checks-status/deepspace2/styleframe/master?label=Tests&logo=github&style=plastic)](https://github.com/DeepSpace2/StyleFrame/actions/workflows/test.yml?query=branch%3Amaster)
[![PyPI](https://img.shields.io/pypi/v/styleframe.svg?style=plastic)](https://pypi.org/project/styleframe/)
![PyPI](https://img.shields.io/pypi/pyversions/StyleFrame.svg?style=plastic)
[![Downloads](http://pepy.tech/badge/styleframe)](https://pepy.tech/project/styleframe)
[![Documentation Status](https://readthedocs.org/projects/styleframe/badge/?version=latest&style=plastic)](https://styleframe.readthedocs.io/en/latest/?badge=latest)

# StyleFrame
_Exporting DataFrames to a styled Excel file has never been so easy_

A library that wraps pandas and openpyxl and allows easy styling of dataframes in Excel.

- [Documentation](http://styleframe.readthedocs.org/en/latest/)
- [Changelog](CHANGELOG.md)
- [How to contribute](.github/CONTRIBUTING.md)

---

## Contents
1. [Rationale](#rationale)
2. [Installation](#installation)
3. [Basics](#basics)
4. [Usage Examples](#usage-examples)    
&nbsp;&nbsp;&nbsp;&nbsp;- [Simple Example](#simple-example)    
&nbsp;&nbsp;&nbsp;&nbsp;- [Advance Example](#advance-example)   
5. [Commandline Interface](#commandline-interface)
6. [Buy me a Pizza](#buy-me-a-pizza)


## Rationale

Pandas's DataFrame is great.   
Dealing with a lot of data is not easy and DataFrame helps us to manage it in the best way possible.   
 
There are many ways to present the output and one of them is excel files.   
Excel files are easy to understand, can be viewed offline, can be sent over the email
and a large percentage of the population familiar with it.   
That is why many times we would choose excel files as our output.   

StyleFrame package allows us to design the excel file on the data in a similar way to DataFrame api.   
It saves us the trouble of working with excel workbook and the suffering of trying to match it with the data stored in our DataFrame.


## Installation
```
$ pip install styleframe
```

## Basics

* ***Styler***:
The `Styler` class represents a style of a cell.   

* ***utils***:
The `utils` module contains helper classes for frequently used styling elements,
such as number and date formats, colors and border types.

* ***Container***: 
The `Container` class represents a cell, a value/style pair.

* ***StyleFrame***:
The `StyleFrame` is the main interaction point you will have. 
It wraps the `DataFrame` object you will be styling.

## Usage Examples

### Simple Example

```python
import pandas as pd
from styleframe import StyleFrame, Styler, utils   
   
df = pd.DataFrame({
    'Time': [1.496728e+09, 1.496728e+09, 1.496728e+09, 1.496728e+09, 1.496728e+09],
    'Expect': ['Hey', 'how', 'are', 'you', 'today?'],
    'Actual': ['Hello', 'how', 'are', 'u', 'today?'],
    'Pass/Fail': ['Failed', 'Passed', 'Passed', 'Failed', 'Passed']
    },
    columns=['Time', 'Expect', 'Actual', 'Pass/Fail'])
   
"""Our DataFrame looks like this:

           Time  Expect  Actual Pass/Fail
0  1.496728e+09     Hey   Hello    Failed
1  1.496728e+09     how     how    Passed
2  1.496728e+09     are     are    Passed
3  1.496728e+09     you       u    Failed
4  1.496728e+09  today?  today?    Passed

"""
   
# Create StyleFrame object that wrap our DataFrame and assign default style.
default_style = Styler(font=utils.fonts.aharoni, font_size=14)
sf = StyleFrame(df, styler_obj=default_style)
   
# Style the headers of the table
header_style = Styler(bold=True, font_size=18)
sf.apply_headers_style(styler_obj=header_style)
   
# Set the background color to green where the test marked as 'passed'
passed_style = Styler(bg_color=utils.colors.green, font_color=utils.colors.white)
sf.apply_style_by_indexes(indexes_to_style=sf[sf['Pass/Fail'] == 'Passed'],
                          cols_to_style='Pass/Fail',
                          styler_obj=passed_style,
                          overwrite_default_style=False)
   
# Set the background color to red where the test marked as 'failed'
failed_style = Styler(bg_color=utils.colors.red, font_color=utils.colors.white)
sf.apply_style_by_indexes(indexes_to_style=sf[sf['Pass/Fail'] == 'Failed'],
                          cols_to_style='Pass/Fail',
                          styler_obj=failed_style,
                          overwrite_default_style=False)
   
# Change the columns width and the rows height
sf.set_column_width(columns=sf.columns, width=20)
sf.set_row_height(rows=sf.row_indexes, height=25)

writer = sf.to_excel('output.xlsx',
                     # Add filters in row 0 to each column.
                     row_to_add_filters=0, 
                     # Freeze the columns before column 'A' (=None)
                     # and rows above '2' (=1).
                     columns_and_rows_to_freeze='A2')

writer.close()
```    
The final output saved under output.xlsx:    
![Example 1](readme-images/example1.PNG?raw=true)


### Advance Example

First, let us create a DataFrame that contains data we would like to export to an .xlsx file 
```python
from datetime import date
import pandas as pd
   
   
columns = ['Date', 'Col A', 'Col B', 'Col C', 'Percentage']
df = pd.DataFrame(data={'Date': [date(1995, 9, 5), date(1947, 11, 29), date(2000, 1, 15)],
                        'Col A': [1, 2004, -3],
                        'Col B': [15, 3, 116],
                        'Col C': [33, -6, 9],
                        'Percentage': [0.113, 0.504, 0.005]},
                  columns=columns)

only_values_df = df[columns[1:-1]]

rows_max_value = only_values_df.idxmax(axis=1)

df['Sum'] = only_values_df.sum(axis=1)
df['Mean'] = only_values_df.mean(axis=1)

"""Our DataFrame looks like this:

         Date  Col A  Col B  Col C  Percentage   Sum        Mean
0  1995-09-05      1     15     33       0.113    49   16.333333
1  1947-11-29   2004      3     -6       0.504  2001  667.000000
2  2000-01-15     -3    116      9       0.005   122   40.666667

"""
```

Now, once we have the DataFrame ready, lets create a StyleFrame object
```python
from styleframe import StyleFrame

sf = StyleFrame(df)
# it is also possible to directly initiate StyleFrame
sf = StyleFrame({'Date': [date(1995, 9, 5), date(1947, 11, 29), date(2000, 1, 15)],
                 'Col A': [1, 2004, -3],
                 'Col B': [15, 3, 116],
                 'Col C': [33, -6, 9],
                 'Percentage': [0.113, 0.504, 0.005],
                 'Sum': [49, 2001, 122],
                 'Mean': [16.333333, 667.000000, 40.666667]})
```

The StyleFrame object will auto-adjust the columns width and the rows height
but they can be changed manually
```python
sf.set_column_width_dict(col_width_dict={
    ('Col A', 'Col B', 'Col C'): 15.3,
    ('Sum', 'Mean'): 30,
    ('Percentage', ): 12
})

# excel rows starts from 1
# row number 1 is the headers
# len of StyleFrame (same as DataFrame) does not count the headers row
all_rows = sf.row_indexes
sf.set_row_height_dict(row_height_dict={
    all_rows[0]: 45,  # headers row
    all_rows[1:]: 25
})
```

Applying number formats
```python
from styleframe import Styler, utils


sf.apply_column_style(cols_to_style='Date',
                      styler_obj=Styler(date_format=utils.number_formats.date,
                                        font=utils.fonts.calibri,
                                        bold=True))

sf.apply_column_style(cols_to_style='Percentage',
                      styler_obj=Styler(number_format=utils.number_formats.percent))

sf.apply_column_style(cols_to_style=['Col A', 'Col B', 'Col C'],
                      styler_obj=Styler(number_format=utils.number_formats.thousands_comma_sep))                     
```

Next, let's change the background color of the maximum values to red and the font to white  
we will also protect those cells and prevent the ability to change their value
```python
style = Styler(bg_color=utils.colors.red,
               bold=True,
               font_color=utils.colors.white,
               protection=True,
               underline=utils.underline.double,
               number_format=utils.number_formats.thousands_comma_sep).to_openpyxl_style()
        
for row_index, col_name in rows_max_value.iteritems():
    sf[col_name][row_index].style = style
```

And change the font and the font size of Sum and Mean columns
```python
sf.apply_column_style(cols_to_style=['Sum', 'Mean'],
                      style_header=True,
                      styler_obj=Styler(font_color='#40B5BF',
                                        font_size=18,
                                        bold=True))
```

Change the background of all rows where the date is after 14/1/2000 to green
```python                 
sf.apply_style_by_indexes(indexes_to_style=sf[sf['Date'] > date(2000, 1, 14)],
                          cols_to_style='Date',
                          styler_obj=Styler(bg_color=utils.colors.green,
                                            date_format=utils.number_formats.date,
                                            bold=True))
```

Finally, let's export to Excel but not before we use more of StyleFrame's features:
- Change the page writing side
- Freeze rows and columns
- Add filters to headers

```python
ew = StyleFrame.ExcelWriter('sf tutorial.xlsx')
sf.to_excel(excel_writer=ew,
            sheet_name='1',
            right_to_left=False,
            columns_and_rows_to_freeze='B2', # will freeze the rows above 2 (=row 1 only) and columns that before column 'B' (=col A only)
            row_to_add_filters=0,
            allow_protection=True)
```

Adding another excel sheet
```python
other_sheet_sf = StyleFrame({'Dates': [date(2016, 10, 20), date(2016, 10, 21), date(2016, 10, 22)]},
                            styler_obj=Styler(date_format=utils.number_formats.date))
                            
other_sheet_sf.to_excel(excel_writer=ew, sheet_name='2')
```

Don't forget to save
```python
ew.close()
```

**_the result:_**
Sheet 1:

![Example 2 sheet 1](readme-images/example2-sheet-1.PNG?raw=true)

Sheet 2:

![Example 2 sheet 2](readme-images/example2-sheet-2.PNG?raw=true)

Note that Sheet 1 is protected:

![Example 2 sheets](readme-images/example2-sheets.PNG?raw=true)

## Commandline Interface
#### General Information

Starting with version 1.1 StyleFrame offers a commandline interface
that lets you create an xlsx file from a json file.

#### Usage

`-v` Displays the installed versions of StyleFrame and its dependencies.

`--json_path` Path to the json file.

`--output_path` Path to the output xlsx file. If not provided defaults to `output.xlsx`.

***Usage Examples:***

`$ styleframe --json_path data.json --output_path data.xlsx`

#### JSON Format

The input JSON should be thought of as an hierarchy of predefined entities,
some of which correspond to a Python class used by StyleFrame.
The top-most level should be a list of `sheet` entities (see below).

An example JSON is available <a href="examples/json_example.json" target="_blank">here</a>.

* ***style:***

Corresponds to: Styler class.

This entity uses the arguments of `Styler.__init__()` as keys.
Any missing keys in the JSON will be given the same default values.

`"style": {"bg_color": "yellow", "bold": true}`

* ***cell***

This entity represents a single cell in the sheet.

Required keys:

`"value"` - The cell's value.

Optional keys:

`"style"` - The `style` entity for this cell. 
If not provided, the `style` provided to the `coloumn` entity will be used.
If that was not provided as well, the default `Styler.__init__()` values will be used.  

`{"value": 42, "style": {"border": "double"}}`

* ***column***

This entity represents a column in the sheet.

Required keys:

`"col_name"` - The column name.

`"cells"` - A list of `cell` entities.

Optional keys:

`"style"` - A style used for the entire column. If not provided the default `Styler.__init__()` values will be used. 

`"width"` - The column's width. If not provided Excel's default column width will be used.

* ***sheet***

This entity represents the whole sheet.

Required keys:

`"sheet_name"` - The sheet's name.

`"columns"` - A list of `column` entities.

Optional keys:

`"default_styles"` - A JSON object with items as keys and `style` entities as values.
Currently only `headers` is supported as an item.

`"default_styles": {"headers": {"bg_color": "blue"}}`
 
`"row_heights"` - A JSON object with rows indexes as keys and heights as value.

`"extra_features"` - A JSON that contains the same arguments as the
`to_excel` method, such as `"row_to_add_filters"`, `"columns_and_rows_to_freeze"`,
`"columns_to_hide"`, `"right_to_left"` and `"allow_protection"`. 
You can also use other arguments that Pandas' `"to_excel"` accepts.

Refer to the <a href="http://styleframe.readthedocs.org/en/latest/" target="_blank">documentation</a>
for more information.

## Buy me a Pizza

You made it all the way down here, cool! If you love StyleFrame and want to show your appreciation, consider buying me a slice of Pizza (or even just drop a comment). Anything will be greatly appreciated 🙂

<a href="https://www.buymeacoffee.com/deepspace"><img src="https://img.buymeacoffee.com/button-api/?text=Buy me a Pizza&emoji=🍕&slug=deepspace&button_colour=5F7FFF&font_colour=ffffff&font_family=Comic&outline_colour=000000&coffee_colour=FFDD00"></a>


CHANGELOG.md
#### 4.2
* **Added Python 3.10 support**
* Added ability to set individual borders' type via the `border_type` argument
  when creating a `Styler` object
* Fixes [GitHub issue #108](https://github.com/DeepSpace2/StyleFrame/issues/108) - Styling and exporting a dataframe that
  contains a column called "index"
* Fixes error when attempting to use `best_fit` argument in `StyleFrame.to_excel`
  on an empty dataframe [GitHub PR #157](https://github.com/DeepSpace2/StyleFrame/pull/157)

#### 4.1
* Added `strikethrough` and `italic` to `Styler`
* Raising `TypeError` in `to_excel` if trying to use a non-openpyxl engine
* Exposing `Exposing StyleFrame.ExcelWriter` and `StyleFrame.read_excel` to the module-level (`styleframe.ExcelWriter`
  and `styleframe.read_excel`) to more closely mimic pandas' API
* Fixes [GitHub issue #124](https://github.com/DeepSpace2/StyleFrame/issues/124) - Improved calculation of color luminosity from tint

#### 4.0
* **Removed Python 3.4 support**
* **Removed Python 3.5 support**
* **Added Python 3.9 support**
* Allowing `StyleFrame.ExcelWriter` to accept any argument (except for `engine`) that `pandas.ExcelWriter` accepts
* Allowing customizing formats of `date`, `time` and `datetime` objects when creating `Styler` instances
* Fixed rows mis-alignment issue when calling `to_excel` with `header=False` ([GitHub issue #88](https://github.com/DeepSpace2/StyleFrame/issues/88))
* `read_excel` does not accept `sheetname` argument anymore (was deprecated since version 1.6). Use `sheet_name` instead.

#### 3.0.6
* Fixes [GitHub issue #94](https://github.com/DeepSpace2/StyleFrame/issues/94) - Passing `border_type=utils.borders.default_grid` to `Styler`

#### 3.0.5
* Fixes [GitHub issue #81](https://github.com/DeepSpace2/StyleFrame/issues/81) - read excel as template headers height

#### 3.0.4
* Fixed style "shifts" when using `read_style=True` and `header=None` with `StyleFrame.read_excel`. Fixes [GitHub issue #80](https://github.com/DeepSpace2/StyleFrame/issues/80) 

#### 3.0.3
No longer relying on openpyxl's colors definition. Related to [GitHub issue #73](https://github.com/DeepSpace2/StyleFrame/issues/73)

#### 3.0.2
Hotfix release - setting maximum versions for dependencies. Related to [GitHub issue #73](https://github.com/DeepSpace2/StyleFrame/issues/73)

#### 3.0.1
* **Removed Python 2.7 support**
* **Added Python 3.8 support**
* Renamed package name to `styleframe` (all lowercase) in accordance of PEP8
* Added `.style` accessor. This allows for easy selection/indexing based on style, for example:
  `sf.loc[sf['col_name'].style.bg_color == utils.colors.yellow]`
  
  or
  
  `sf.loc[~sf['col_name'].style.bold]`
  
* Added `default_grid` to `utils.borders` to allow usage of the default spreadsheet grid
* Added `read_excel_as_template` method
* Fixed a bug that prevented saving if `read_excel` was used with `use_openpxl_style=True`, see [GitHub issue #67](https://github.com/DeepSpace2/StyleFrame/issues/67)
* Allowing usage of pathlib.Path in `to_excel`, see [GitHub issue #69](https://github.com/DeepSpace2/StyleFrame/issues/69) 
* Added ability to execute the tests from the commandline: `styleframe --test`

#### 2.0.5
* `style_alternate_rows` can accept all arguments that `apply_style_by_indexes` accepts as kwargs.
* Added `cols_to_style` argument to `apply_headers_style`

#### 2.0.4
* Fixed a bug that caused `apply_style_by_indexes` not to work in case the dataframe had non-integer indexes in some cases
* Added support for text rotation

#### 2.0.3
* Fixing pandas dependency for different Python versions, related to #52, #53

#### 2.0.2
* Fixed a "'column' is out of columns range" error when settings a column's width if dataframe has more than 26 columns. 

#### 2.0.1
* Hotfix: Fixed typo in setup.py

#### 2.0
* **Change supported versions of openpyxl to >= 2.5**
* Changed `use_openpyxl_styles` argument default value to `False` in `StyleFrame.read_excel`.

#### 1.6.2
* Fixed an issue that prevented reading files which used certain theme colors.

#### 1.6.1
* Reading columns' widths and rows' heights when passing `read_style=True` to `read_excel`.
* Added support for named indexes.
* Added `style_index_header` argument to `apply_headers_style`.
* Added support for pandas <= 0.23.4

#### 1.6
* **Added Python 3.7 support**
* Added support for pandas <= 0.23.1
* Added dt and str accessors to Series
* Added support for passing an integer (sheet index) as `sheet_name` to `StyleFrame.read_excel`
* `StyleFrame.read_excel` `sheetname` argument changed to `sheet_name`. Using `sheetname` is still allowed but will
  show deprecation warning.
* Added `Styler.combine` method.
* Added `utils.number_formats.decimal_with_num_of_digits` method.
* Added `overwrite_default_style` argument to `StyleFrame` methods `apply_style_by_indexes` and `apply_column_style`


#### 1.5.1
* Fixed a bug where `read_excel` will fail when using `read_style=True` in cases
  where specific themes are used (See [GitHub issue #37](https://github.com/DeepSpace2/StyleFrame/issues/37)).

#### 1.5
* Added `complement_style` and `complement_height` arguments to `StyleFrame.apply_style_by_indexes`
* Added support for comments
* Renamed `Styler.create_style` method to `to_openpyxl_style` (`create_style` is still available for backward compatibility)
* Fixed a bug not allowing to access StyleFrame columns by dot notation in case a column name is a number

#### 1.4
* **No longer supporting Python 3.3**
* StyleFrame objects no longer expose .ix method as it is deprecated since pandas 0.20. Use .loc or .iloc instead
* Added ability to access StyleFrame columns as attributes (eg `sf.column_a`)
* Added conditional formatting
* Added `best_fit` to `to_excel` method
* Added support for pandas <= 0.22.0
* Added support for theme colors when reading styles from Excel sheets
* Added option to use `Styler` objects when reading styles from Excel sheets
* Using a JSON schema to validate json from command line

styleframe/utils.py
import re

from typing import Union, Match, Optional, Any

HEX_REGEX = re.compile(r'^([A-Fa-f0-9]{8}|[A-Fa-f0-9]{6})$')


def is_hex_color_string(hex_string: Optional[str]) -> Union[Match[Any], bool]:
    return HEX_REGEX.match(hex_string) if hex_string else False


class BaseDefClass:
    @classmethod
    def get(cls, key, default=None):
        return cls.__dict__.get(key, default)


# The following classes names violate PEP8 for the sake of keeping backwards compatibility, at least for the meantime

class number_formats(BaseDefClass):
    """
    :cvar str general: 'General'
    :cvar str general_integer: '0'
    :cvar str general_float: '0.00'
    :cvar str percent: '0.0%'
    :cvar str thousands_comma_sep: '#,##0'
    :cvar str date: 'DD/MM/YY'
    :cvar str time_24_hours: 'HH:MM'
    :cvar str time_24_hours_with_seconds: 'HH:MM:SS'
    :cvar str time_12_hours: 'h:MM AM/PM'
    :cvar str time_12_hours_with_seconds: 'h:MM:SS AM/PM'
    :cvar str date_time: 'DD/MM/YY HH:MM'
    :cvar str date_time_with_seconds: 'DD/MM/YY HH:MM:SS'
    """

    general = 'General'
    general_integer = '0'
    general_float = '0.00'
    percent = '0.0%'
    thousands_comma_sep = '#,##0'
    date = 'DD/MM/YY'
    time_24_hours = 'HH:MM'
    time_24_hours_with_seconds = 'HH:MM:SS'
    time_12_hours = 'h:MM AM/PM'
    time_12_hours_with_seconds = 'h:MM:SS AM/PM'
    date_time = 'DD/MM/YY HH:MM'
    date_time_with_seconds = 'DD/MM/YY HH:MM:SS'

    default_date_format = date
    default_time_format = time_24_hours
    default_date_time_format = date_time

    @staticmethod
    def decimal_with_num_of_digits(num_of_digits: int) -> str:
        """
        .. versionadded:: 1.6

        :param int num_of_digits: Number of digits after the decimal point
        :return: A format string that represents a floating point number with the provided number of digits after the
            decimal point.

            For example, ``utils.number_formats.decimal_with_num_of_digits(2)`` will return ``'0.00'``
        :rtype: str
        """
        return '0.{}'.format('0' * num_of_digits)


class colors(BaseDefClass):
    """
    :cvar str white: '00FFFFFF'
    :cvar str blue: '000000FF'
    :cvar str dark_blue: '00000080'
    :cvar str yellow: '00FFFF00'
    :cvar str dark_yellow: '00808000'
    :cvar str green: '0000FF00'
    :cvar str dark_green: '00008000'
    :cvar str black: '00000000'
    :cvar str red: '00FF0000'
    :cvar str dark_red: '00800000'
    :cvar str purple: '800080'
    :cvar str grey: 'D3D3D3'
    """
    
    white = '00FFFFFF'
    blue = '000000FF'
    dark_blue = '00000080'
    yellow = '00FFFF00'
    dark_yellow = '00808000'
    green = '0000FF00'
    dark_green = '00008000'
    black = '00000000'
    red = '00FF0000'
    dark_red = '00800000'
    purple = '800080'
    grey = 'D3D3D3'


class fonts(BaseDefClass):
    """
    .. versionadded:: 1.1

    :cvar str aegean: 'Aegean'
    :cvar str aegyptus: 'Aegyptus'
    :cvar str aharoni: 'Aharoni CLM'
    :cvar str anaktoria: 'Anaktoria'
    :cvar str analecta: 'Analecta'
    :cvar str anatolian: 'Anatolian'
    :cvar str arial: 'Arial'
    :cvar str calibri: 'Calibri'
    :cvar str david: 'David CLM'
    :cvar str dejavu_sans: 'DejaVu Sans'
    :cvar str ellinia: 'Ellinia CLM'
    """

    aegean = 'Aegean'
    aegyptus = 'Aegyptus'
    aharoni = 'Aharoni CLM'
    anaktoria = 'Anaktoria'
    analecta = 'Analecta'
    anatolian = 'Anatolian'
    arial = 'Arial'
    calibri = 'Calibri'
    david = 'David CLM'
    dejavu_sans = 'DejaVu Sans'
    ellinia = 'Ellinia CLM'


class borders(BaseDefClass):
    """
    :cvar str dash_dot: 'dashDot'
    :cvar str dash_dot_dot: 'dashDotDot'
    :cvar str dashed: 'dashed'
    :cvar str default_grid: 'default_grid'
    :cvar str dotted: 'dotted'
    :cvar str double: 'double'
    :cvar str hair: 'hair'
    :cvar str medium: 'medium'
    :cvar str medium_dash_dot: 'mediumDashDot'
    :cvar str medium_dash_dot_dot: 'mediumDashDotDot'
    :cvar str medium_dashed: 'mediumDashed'
    :cvar str slant_dash_dot: 'slantDashDot'
    :cvar str thick: 'thick'
    :cvar str thin: 'thin'
    """

    dash_dot = 'dashDot'
    dash_dot_dot = 'dashDotDot'
    dashed = 'dashed'
    default_grid = 'default_grid'
    dotted = 'dotted'
    double = 'double'
    hair = 'hair'
    medium = 'medium'
    medium_dash_dot = 'mediumDashDot'
    medium_dash_dot_dot = 'mediumDashDotDot'
    medium_dashed = 'mediumDashed'
    slant_dash_dot = 'slantDashDot'
    thick = 'thick'
    thin = 'thin'


class border_locations(BaseDefClass):
    """
    .. versionadded:: 4.2

    :cvar str top: 'top'
    :cvar str bottom: 'bottom'
    :cvar str left: 'left'
    :cvar str right: 'right'
    """

    top = 'top'
    bottom = 'bottom'
    left = 'left'
    right = 'right'


class horizontal_alignments(BaseDefClass):
    """
    :cvar str general: 'general'
    :cvar str left: 'left'
    :cvar str center: 'center'
    :cvar str right: 'right'
    :cvar str fill: 'fill'
    :cvar str justify: 'justify'
    :cvar str center_continuous: 'centerContinuous'
    :cvar str distributed: 'distributed'
    """
    
    general = 'general'
    left = 'left'
    center = 'center'
    right = 'right'
    fill = 'fill'
    justify = 'justify'
    center_continuous = 'centerContinuous'
    distributed = 'distributed'


class vertical_alignments(BaseDefClass):
    """
    :cvar str top: 'top'
    :cvar str center: 'center'
    :cvar str bottom: 'bottom'
    :cvar str justify: 'justify'
    :cvar str distributed: 'distributed'
    """
    
    top = 'top'
    center = 'center'
    bottom = 'bottom'
    justify = 'justify'
    distributed = 'distributed'


class underline(BaseDefClass):
    """
    :cvar str single: 'single'
    :cvar str double: 'double'
    """
    
    single = 'single'
    double = 'double'


class fill_pattern_types(BaseDefClass):
    """
    .. versionadded:: 1.2

    :cvar str solid: 'solid'
    :cvar str dark_down: 'darkDown'
    :cvar str dark_gray: 'darkGray'
    :cvar str dark_grid: 'darkGrid'
    :cvar str dark_horizontal: 'darkHorizontal'
    :cvar str dark_trellis: 'darkTrellis'
    :cvar str dark_up: 'darkUp'
    :cvar str dark_vertical: 'darkVertical'
    :cvar str gray0625: 'gray0625'
    :cvar str gray125: 'gray125'
    :cvar str light_down: 'lightDown'
    :cvar str light_gray: 'lightGray'
    :cvar str light_grid: 'lightGrid'
    :cvar str light_horizontal: 'lightHorizontal'
    :cvar str light_trellis: 'lightTrellis'
    :cvar str light_up: 'lightUp'
    :cvar str light_vertical: 'lightVertical'
    :cvar str medium_gray: 'mediumGray'
    """
    
    solid = 'solid'
    dark_down = 'darkDown'
    dark_gray = 'darkGray'
    dark_grid = 'darkGrid'
    dark_horizontal = 'darkHorizontal'
    dark_trellis = 'darkTrellis'
    dark_up = 'darkUp'
    dark_vertical = 'darkVertical'
    gray0625 = 'gray0625'
    gray125 = 'gray125'
    light_down = 'lightDown'
    light_gray = 'lightGray'
    light_grid = 'lightGrid'
    light_horizontal = 'lightHorizontal'
    light_trellis = 'lightTrellis'
    light_up = 'lightUp'
    light_vertical = 'lightVertical'
    medium_gray = 'mediumGray'


class conditional_formatting_types(BaseDefClass):
    """
    :cvar str num: 'num'
    :cvar str percent: 'percent'
    :cvar str max: 'max'
    :cvar str min: 'min'
    :cvar str formula: 'formula'
    :cvar str percentile: 'percentile'
    """
    
    num = 'num'
    percent = 'percent'
    max = 'max'
    min = 'min'
    formula = 'formula'
    percentile = 'percentile'
